### ISSUE TITLE
[Security] Disable introspection in `test` and `live` modes

### ISSUE THREAD

**Issue by madmatt** [silverstripe/silverstripe-graphql#420]
I'm a GraphQL newbie, so I might be missing something here, but is there any reason why you might want introspection enabled in production for the `/admin/graphql` admin GraphQL endpoint? During development it's obviously helpful, and if you're shipping frontend, non-CMS GraphQL APIs you might want this, but by default I think it makes sense to disable this for the /admin/graphql endpoint.

Thoughts? Am I way off base with this one?

**Comment by unclecheese** [comment]
Relevant: https://github.com/silverstripe/silverstripe-graphql/pull/400

We're also tracking this as a security issue. Gets a very low score because by default we only expose the `admin` schema, which is already behind auth.

Still, it's a footgun, and we should provide sensible defaults.

**Comment by maxime-rainville** [comment]
[OWASP GraphQL cheat sheet](https://cheatsheetseries.owasp.org/cheatsheets/GraphQL_Cheat_Sheet.html) treats it more as a best practice.

Just looking at the CMS graphql schema, that's entirely public. All you would need to do to find it is installed the CMS in and look at it.

If you had your own schema and we're developing an application targeted at other devs, there's a perfectly valid use case where you would want to leave introspection on to make it easy for them to debug their app.

Can't imagine we want to backport this to GraphQL v3. 

It's just a question of adding an option to disable introspection to GraphQL v4. The key question now is do we want introspection ON or OFF by default on TEST/LIVE?

**Comment by baukezwaan** [comment]
> It's just a question of adding an option to disable introspection to GraphQL v4. The key question now is do we want introspection ON or OFF by default on TEST/LIVE?

Best practise would be to turn it off on production- and test-environments. And you can leave it open on dev-environments.

.
### Disables schema introspection outside of dev

For anyone bumping into this thread, looking for _how_ to disable introspection on GraphQL 4v. This is what you can do:

Create a middleware PHP-file:
```
    namespace MyApp\GraphQL\Middleware;

    class IntrospectionMiddleware implements QueryMiddleware
    {
        public function process(Schema $schema, string $query, array $context, array $vars, callable $next)
        {
            DocumentValidator::addRule(new DisableIntrospection());
            
            return $next($schema, $query, $context, $vars);
        }
    }
```

And a yml-config file, to set the middelware for everywhere except dev:
```
---
Name: 'app-graphql-middlewares'
Except:
  environment: dev
---
SilverStripe\Core\Injector\Injector:
  # use SilverStripe\GraphQL\QueryHandler\QueryHandlerInterface.default or .admin if you just want to apply to one schema
  SilverStripe\GraphQL\QueryHandler\QueryHandlerInterface:
    properties:
      Middlewares:
        disableIntrospection: '%$MyApp\GraphQL\Middleware\IntrospectionMiddleware'
```

Thanks to @unclecheese for the help on this

**Comment by sunnysideup** [comment]
@baukezwaan  - does this work for you? I dont get any errors, yet to test the outcome.

**Comment by baukezwaan** [comment]
Yes, it worked for us. I do remember having some issues locally with cache.

**Comment by MLindenhofer** [comment]
Is there a way to disable it in CMS5 too? 

I tried it like this but I can still do the following query at the admin schema:

```
class IntrospectionMiddleware implements QueryMiddleware
{
    public function process(Schema $schema, string $query, array $context, array $vars, callable $next)
    {
        DocumentValidator::addRule(new DisableIntrospection(false));

        return $next($schema, $query, $context, $vars);
    }
}
```

Query:
`{
  __schema{
    types {
      name
    }
  }
}`

**Comment by guywatson** [comment]
@MLindenhofer A bit late now, but it needs to be

`DocumentValidator::addRule(new DisableIntrospection(DisableIntrospection::ENABLED));`

that worked for me in SS5

**Comment by sunnysideup** [comment]
If this issue is open but there is a fix then can we put the fix into the module?
